src/postamble.js

```diff
@@ -175,7 +175,11 @@ Module['callMain'] = Module.callMain = function callMain(args) {
 #endif
       return;
     } else {
-      if (e && typeof e === 'object' && e.stack) Module.printErr('exception thrown: ' + [e, e.stack]);
+      var hasStack = e && typeof e === 'object' && e.stack;
+      if (hasStack) Module.printErr('exception thrown: ' + [e, e.stack]);
+      // In a shell environment, callMain is ran from a promise context. Exceptions thrown from
+      // promises don't set the shell's exit code, so call quit here to force it to be non-zero.
+      quitIfShell(1, hasStack ? '' : ('exception thrown: ' + e));
```

Comment:
how about calling `exit` here?

 * It would have the `exit` logic of handling `NO_EXIT_RUNTIME`etc.
 * It would avoid having shell-specific logic here and in other places.

In fact we could unify the `exit` call higher up, which returns ret - we could set the returncode from main(), or if an exception is thrown, set it to 1 here. Then in a `finally` clause we call `exit`.